### PR TITLE
Stop `Balances` pallet erroneously double incrementing and decrementing consumers

### PR DESCRIPTION
**Pull request by liamaharon**
Closes https://github.com/paritytech/polkadot-sdk/issues/1970

Follow up issue to tackle, once the erroneous double incrementing/decrementing has stopped: https://github.com/paritytech/polkadot-sdk/issues/2037

### CODE REVIEW

**Comment by kianenigma** [comment]
> Perhaps we could write a script that scrapes extrinsics for for accounts that had their consumers erroneously incremented and generates a migration to decrement them?

First glance, this will be a race that you're alway behind, as more accounts can get their consumers incremented by mistake in the timespan while we wait for the migration to execute.

**Comment by bkchr** [comment]
We should first deploy the fix and then let the fix come in a later runtime upgrade.

File: substrate/frame/balances/src/lib.rs [hunk 1]
```diff
@@ -1005,26 +1005,9 @@ pub mod pallet {
...
-					// No providers for the locks. This is impossible under normal circumstances
-					// since the funds that are under the lock will themselves be stored in the
-					// account and therefore will need a reference.
-					log::warn!(
-						target: LOG_TARGET,
-						"Warning: Attempt to introduce lock consumer reference, yet no providers. \
-						This is unexpected but should be safe."
-					);
-				}
+			match locks.is_empty() {
```
Comment (franciscoaguirre) [inline substrate/frame/balances/src/lib.rs:1008]: This could just be an `if`

File: substrate/frame/balances/src/tests/currency_tests.rs [hunk 2]
```diff
@@ -144,7 +144,9 @@ fn lock_removal_should_work() {
 		.monied(true)
 		.build_and_execute_with(|| {
 			Balances::set_lock(ID_1, &1, u64::MAX, WithdrawReasons::all());
+			assert_eq!(System::consumers(&1), 1);
 			Balances::remove_lock(ID_1, &1);
+			assert_eq!(System::consumers(&1), 0);
```
Comment (franciscoaguirre) [inline substrate/frame/balances/src/tests/currency_tests.rs:149]: Is it possible for the balances pallet to increase consumers more than 1? We could add more tests in that case

**Review by franciscoaguirre — APPROVED** [review]

File: substrate/frame/balances/src/lib.rs (hunk 1 above)
Comment (juangirini) [reply substrate/frame/balances/src/lib.rs:1008]: It was an `if` indeed, I do like the readability of `match`

**Review by juangirini — APPROVED** [review]

**Review by kianenigma — APPROVED** [review]

File: substrate/frame/balances/src/tests/currency_tests.rs (hunk 2 above)
Comment (liamaharon) [reply substrate/frame/balances/src/tests/currency_tests.rs:149]: I don't think it's possible.

**Comment by ggwpez** [comment]
What do do with chains that got fucked by this before the fix applied?  
eg https://substrate.stackexchange.com/questions/11166

**Comment by liamaharon** [comment]
> What to do with chains that got fucked by this before the fix applied? eg https://substrate.stackexchange.com/questions/11166

Need to retroactively fix it https://github.com/paritytech/polkadot-sdk/issues/2037

**Comment by ggwpez** [comment]
Screenshot 2024-03-21 at 18 44 24

I think this fixed the `(F, T)` case but broke the `(T, T)` case.

**Pull request merged**